services-directions-models/src/main/java/com/mapbox/api/directions/v5/utils/FormatUtils.java

```diff
@@ -0,0 +1,211 @@
+   * @return {@link String}
+   */
+  public static String join(CharSequence delimiter, Object[] tokens) {
+    if (tokens == null || tokens.length < 1) {
+      return null;
+    }
+
+    StringBuilder sb = new StringBuilder();
+    boolean firstTime = true;
+    for (Object token : tokens) {
+      if (firstTime) {
+        firstTime = false;
+      } else {
+        sb.append(delimiter);
+      }
+      sb.append(token);
+    }
+    return sb.toString();
+  }
+
+  /**
+   * Returns a string containing the tokens joined by delimiters.
+   *
+   * @param delimiter the delimiter on which to split.
+   * @param tokens A list of objects to be joined. Strings will be formed from the objects by
+   *               calling object.toString().
+   * @return {@link String}
+   */
+  public static String join(CharSequence delimiter, List<?> tokens) {
+    if (tokens == null || tokens.size() < 1) {
+      return null;
+    }
+    return join(delimiter, tokens.toArray());
+  }
+
+  /**
+   * Useful to remove any trailing zeros and prevent a coordinate being over 7 significant figures.
+   *
+   * @param coordinate a double value representing a coordinate.
+   * @return a formatted string.
+   */
+  public static String formatCoordinate(double coordinate) {
+    DecimalFormat decimalFormat = new DecimalFormat("0.######",
+        new DecimalFormatSymbols(Locale.US));
+    return String.format(Locale.US, "%s",
+        decimalFormat.format(coordinate));
+  }
+
+  /**
+   * Used in various APIs to format the user provided radiuses to a String matching the APIs
+   * format.
+   *
+   * @param radiuses a list of doubles represents the radius values
+   * @return a String ready for being passed into the Retrofit call
+   */
+  public static String formatRadiuses(List<Double> radiuses) {
+    if (radiuses == null || radiuses.size() == 0) {
+      return null;
+    }
+
+    String[] radiusesFormatted = new String[radiuses.size()];
+    for (int i = 0; i < radiuses.size(); i++) {
+      if (radiuses.get(i) == Double.POSITIVE_INFINITY) {
+        radiusesFormatted[i] = "unlimited";
+      } else {
+        radiusesFormatted[i] = String.format(Locale.US, "%s", formatCoordinate(radiuses.get(i)));
+      }
+    }
+    return join(";", radiusesFormatted);
+  }
+
+  /**
+   * Formats the bearing variables from the raw values to a string which can than be used for the
+   * request URL.
+   *
+   * @param bearings a List of list of doubles representing bearing values
+   * @return a string with the bearing values
+   */
+  public static String formatBearings(List<List<Double>> bearings) {
+    if (bearings.isEmpty()) {
+      return null;
+    }
+
+    String[] bearingFormatted = new String[bearings.size()];
+    for (int i = 0; i < bearings.size(); i++) {
+      if (bearings.get(i).size() == 0) {
+        bearingFormatted[i] = "";
+      } else {
+        bearingFormatted[i] = String.format(Locale.US, "%s,%s",
+            formatCoordinate(bearings.get(i).get(0)),
```

Comment:
The nested coordinate lists are nullable if we specify `null` as the bearing for a specific waypoint or origin.

Reply:
Yeah, as mentioned we should double check the formatting methods as there are cases in which `null` is valid and gets formatted to `""`.